### PR TITLE
[ML-248][LinearRegression] Add Linear Regression GPU algorithm

### LINKED ISSUE #248: [LinearRegression] Add Linear Regression GPU algorithm

### PR DESCRIPTION
**Pull request by argentea**
close #248

### CODE REVIEW

File: dev/prepare-build-deps-gpu.sh
Comment (xwu99) [inline dev/prepare-build-deps-gpu.sh:1]: Could you work with @minmingzhu to merge the version changes in build scripts to #259

File: mllib-dal/src/main/native/LinearRegressionDALImpl.cpp [hunk 1]
```diff
@@ -241,3 +260,63 @@ Java_com_intel_oap_mllib_regression_LinearRegressionDALImpl_cLinearRegressionTra
...
+    }
+	*/
+	return (jlong)0;
+}
+#endif
+
+
+#ifdef CPU_GPU_PROFILE
+JNIEXPORT jlong JNICALL
+Java_com_intel_oap_mllib_regression_KMeansDALImpl_cLROneapiCompute(
```
Comment (carsonwang) [inline mllib-dal/src/main/native/LinearRegressionDALImpl.cpp:309]: A lot of code copied from KMeans. Please update and clean up.

File: examples/linear-regression/run.sh
Comment (xwu99) [inline examples/linear-regression/run.sh:1]: Please follow kmeans example, some changes for run.sh and need to create a new script named "run-gpu-standalone.sh".

File: examples/linear-regression/src/main/scala/org/apache/spark/examples/ml/LinearRegressionExample.scala [hunk 2]
```diff
@@ -108,8 +110,16 @@ object LinearRegressionExample {
     println(s"LinearRegressionExample with parameters:\n$params")
 
     val training = spark.read.format("libsvm")
-      .load(params.input).toDF("label", "features")
+      .load(params.input).toDF("label", "features").repartition(4)
```
Comment (xwu99) [inline examples/linear-regression/src/main/scala/org/apache/spark/examples/ml/LinearRegressionExample.scala:113]: don't use repartition in user code, you should check if partition is enough and change in oap mllib code.

File: examples/linear-regression/src/main/scala/org/apache/spark/examples/ml/LinearRegressionExample.scala [hunk 3]
```diff
@@ -108,8 +110,16 @@ object LinearRegressionExample {
...
 
+    training.select("label","features").printSchema()
+    val featuresRDD=training
+      .select("label","features").rdd.map{
+        case Row(label:Double,feature:Vector)=>new LabeledPoint(label,feature.toDense)
```
Comment (xwu99) [inline examples/linear-regression/src/main/scala/org/apache/spark/examples/ml/LinearRegressionExample.scala:118]: pay attention to code style. extra space after ","

File: mllib-dal/src/main/native/LinearRegressionDALImpl.cpp [hunk 4]
```diff
@@ -15,16 +15,36 @@
...
+
+#include <iomanip>
 #include <iostream>
 #include <vector>
 
+#ifdef CPU_GPU_PROFILE
+#include "GPU.h"
+#ifndef ONEDAL_DATA_PARALLEL
+#define ONEDAL_DATA_PARALLEL
+#endif
```
Comment (xwu99) [inline mllib-dal/src/main/native/LinearRegressionDALImpl.cpp:28]: compile error here due to some changes to the include logic, you need to include Common.hpp, pls follow KMeansImpl.cpp

File: mllib-dal/src/main/scala/com/intel/oap/mllib/regression/LinearRegressionDALImpl.scala [hunk 5]
```diff
@@ -49,73 +57,107 @@ class LinearRegressionDALImpl( val fitIntercept: Boolean,
...
+
+      val sparkContext = labeledPoints.sparkSession.sparkContext
+      val useDevice = sparkContext.getConf.get("spark.oap.mllib.device", Utils.DefaultComputeDevice)
+      val computeDevice = Common.ComputeDevice.getDeviceByName(useDevice)
+
+      val kvsIPPort = getOneCCLIPPort(labeledPoints.rdd)
+
+      val labeledPointsTables = if (useDevice == "GPU") {
+          if (OneDAL.isDenseDataset(labeledPoints, featuresCol)) {
+            println(labeledPoints.select(labelCol).count())
```
Comment (xwu99) [inline mllib-dal/src/main/scala/com/intel/oap/mllib/regression/LinearRegressionDALImpl.scala:80]: no debug code please.

File: mllib-dal/src/main/scala/com/intel/oap/mllib/regression/LinearRegressionDALImpl.scala [hunk 6]
```diff
@@ -49,73 +57,107 @@ class LinearRegressionDALImpl( val fitIntercept: Boolean,
...
+      } else {
+          if (OneDAL.isDenseDataset(labeledPoints, featuresCol)) {
+          OneDAL.rddLabeledPointToMergedTables(labeledPoints, labelCol, featuresCol, executorNum)
         } else {
-          Iterator.empty
+          OneDAL.rddLabeledPointToSparseTables(labeledPoints, labelCol, featuresCol, executorNum)
         }
+      }
+
+      println(useDevice)
```
Comment (xwu99) [inline mllib-dal/src/main/scala/com/intel/oap/mllib/regression/LinearRegressionDALImpl.scala:96]: no debug code

File: mllib-dal/src/main/scala/com/intel/oap/mllib/regression/LinearRegressionDALImpl.scala
Comment (xwu99) [inline mllib-dal/src/main/scala/com/intel/oap/mllib/regression/LinearRegressionDALImpl.scala:1]: rename filename to LinearRegressionImpl.cpp

**Review by xwu99 — CHANGES_REQUESTED** [review]

File: mllib-dal/src/main/native/LinearRegressionImpl.cpp [hunk 7]
```diff
@@ -194,50 +207,120 @@ static NumericTablePtr ridge_regression_compute(
     return resultTable;
 }
 
+#ifdef CPU_GPU_PROFILE
+static jlong doLROneAPICompute(JNIEnv *env, jint rankId, jlong pData,
+                               jlong pLabel, jint executorNum,
+                               preview::spmd::communicator<preview::spmd::device_memory_access::usm> comm,
```
Comment (argentea) [inline mllib-dal/src/main/native/LinearRegressionImpl.cpp:213]: variable name

**Pull request merged**